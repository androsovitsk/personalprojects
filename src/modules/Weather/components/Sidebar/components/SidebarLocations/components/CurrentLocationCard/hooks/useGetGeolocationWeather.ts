import useFetchCurrentWeather from '../../../../../hooks/useFetchCurrentWeather'
import useWeatherContext from '../../../../../../../hooks/useWeatherContext'
import { useEffect, useState } from 'react'
import useGeolocation from '../../../../../../../../../ui/hooks/useGeolocation'
import { isNil } from 'ramda'
import ICurrentWeather from '../../../../../../../types/ICurrentWeather'

const useGetGeolocationWeather = () => {
  const [isLoading, setIsLoading] = useState<boolean>(false)
  const [hasError, setHasError] = useState<boolean>(false)
  const [currentWeather, setCurrentWeather] = useState<ICurrentWeather | null>(
    null
  )

  const {
    state: { timestamp }
  } = useWeatherContext()

  const fetchCurrentWeather = useFetchCurrentWeather({
    setIsLoading,
    setHasError,
    setCurrentWeather
  })

  const geolocation = useGeolocation()

  useEffect(() => {
    setIsLoading(geolocation.isLoading)
    setHasError(geolocation.hasError)
  }, [geolocation.isLoading, geolocation.hasError])

  useEffect(() => {
    fetchCurrentWeather(
      {
        lat: geolocation.location?.lat.toString(),
        lon: geolocation.location?.lon.toString()
      },
      !isNil(geolocation.location)
    )
  }, [fetchCurrentWeather, timestamp, geolocation.location])

  return { isLoading, hasError, currentWeather }
}

export default useGetGeolocationWeather
