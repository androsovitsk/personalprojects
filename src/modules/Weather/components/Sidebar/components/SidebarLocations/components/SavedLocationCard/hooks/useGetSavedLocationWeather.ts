import { useEffect, useState } from 'react'
import useFetchCurrentWeather from '../../../../../hooks/useFetchCurrentWeather'
import useWeatherContext from '../../../../../../../hooks/useWeatherContext'
import ICurrentWeather from '../../../../../../../types/ICurrentWeather'

const useGetSavedLocationWeather = (cityName: string) => {
  const [isLoading, setIsLoading] = useState<boolean>(false)
  const [hasError, setHasError] = useState<boolean>(false)
  const [currentWeather, setCurrentWeather] = useState<ICurrentWeather | null>(
    null
  )

  const {
    state: { timestamp }
  } = useWeatherContext()

  const fetchCurrentWeather = useFetchCurrentWeather({
    setIsLoading,
    setHasError,
    setCurrentWeather
  })

  useEffect(() => {
    fetchCurrentWeather({ q: cityName })
  }, [fetchCurrentWeather, timestamp, cityName])

  return { isLoading, hasError, currentWeather }
}

export default useGetSavedLocationWeather
