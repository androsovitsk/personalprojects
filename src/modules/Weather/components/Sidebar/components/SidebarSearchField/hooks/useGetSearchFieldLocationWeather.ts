import useSearchFieldActions from './useSearchFieldActions'
import useWeatherContext from '../../../../../hooks/useWeatherContext'
import useFetchCurrentWeather from '../../../hooks/useFetchCurrentWeather'
import { useEffect, useState } from 'react'
import { isEmpty, isNil } from 'ramda'
import ICurrentWeather from '../../../../../types/ICurrentWeather'

const useGetSearchFieldLocationWeather = (cityName: string | null) => {
  const [isLoading, setIsLoading] = useState<boolean>(false)
  const [hasError, setHasError] = useState<boolean>(false)
  const [currentWeather, setCurrentWeather] = useState<ICurrentWeather | null>(
    null
  )

  const {
    state: { timestamp }
  } = useWeatherContext()

  const fetchCurrentWeather = useFetchCurrentWeather({
    setIsLoading,
    setHasError,
    setCurrentWeather
  })

  useEffect(() => {
    fetchCurrentWeather(
      { q: cityName },
      !isNil(cityName) && !isEmpty(cityName),
      isEmpty(cityName)
    )
  }, [fetchCurrentWeather, timestamp, cityName])

  useSearchFieldActions(isLoading, hasError, currentWeather)
}

export default useGetSearchFieldLocationWeather
