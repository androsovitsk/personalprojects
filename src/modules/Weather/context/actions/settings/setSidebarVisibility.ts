import SettingsReducerAction from '../../types/SettingsReducerAction'

const setSidebarVisibility = (visible: boolean): SettingsReducerAction => ({
  type: 'SET_SIDEBAR_VISIBILITY',
  payload: visible
})

export default setSidebarVisibility
