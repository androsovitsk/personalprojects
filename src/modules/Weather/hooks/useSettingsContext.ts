import { useContext } from 'react'
import SettingsContext from '../context/component/SettingsContext'
import ISettingsContext from '../context/types/ISettingsContext'

const useSettingsContext = () => useContext(SettingsContext) as ISettingsContext

export default useSettingsContext
