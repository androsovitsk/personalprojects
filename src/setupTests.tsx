import '@testing-library/jest-dom'
import React from 'react'
import { render } from '@testing-library/react'
import fetchMock from 'fetch-mock'
import TranslationProvider from './ui/components/TranslationProvider/TranslationProvider'
import LanguageTypes from './ui/components/TranslationProvider/types/LanguageTypes'
import resourceBundles from './locale'
import { createTheme, ThemeProvider } from '@mui/material'

beforeEach(() => {
  fetchMock.reset()
  jest.clearAllMocks()
})

const Wrapper: React.FC<React.PropsWithChildren> = ({ children }) => {
  return (
    <ThemeProvider theme={createTheme({})}>
      <TranslationProvider
        defaultLanguage={LanguageTypes.ENGLISH}
        resourceBundles={resourceBundles}
      >
        {children}
      </TranslationProvider>
    </ThemeProvider>
  )
}

export const renderWithProviders = (element: React.ReactElement) =>
  render(element, {
    wrapper: Wrapper
  })
