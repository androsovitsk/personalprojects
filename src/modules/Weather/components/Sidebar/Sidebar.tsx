import React from 'react'
import Drawer from '@mui/material/Drawer'
import SidebarHeader from './components/SidebarHeader/SidebarHeader'
import SidebarLocations from './components/SidebarLocations/SidebarLocations'
import SidebarSearchField from './components/SidebarSearchField/SidebarSearchField'
import useSettingsContext from '../../hooks/useSettingsContext'
import setSidebarVisibility from '../../context/actions/settings/setSidebarVisibility'

interface ISidebarProps {
  onSettingsButtonClick: () => void
}

const Sidebar: React.FC<ISidebarProps> = ({ onSettingsButtonClick }) => {
  const {
    state: { isSidebarOpen },
    dispatcher
  } = useSettingsContext()

  return (
    <Drawer
      open={isSidebarOpen}
      variant='persistent'
      anchor='left'
      sx={{
        '& .MuiDrawer-paper': (theme) => ({
          [theme.breakpoints.down('md')]: {
            minWidth: `calc(100% - ${theme.spacing(4)})`
          },
          [theme.breakpoints.up('md')]: { width: 360 },
          padding: theme.spacing(2)
        })
      }}
    >
      <SidebarHeader
        onMenuButtonClick={() => dispatcher(setSidebarVisibility(false))}
        onSettingsButtonClick={onSettingsButtonClick}
      />
      <SidebarSearchField />
      <SidebarLocations />
    </Drawer>
  )
}

Sidebar.displayName = 'Sidebar'

export default Sidebar
