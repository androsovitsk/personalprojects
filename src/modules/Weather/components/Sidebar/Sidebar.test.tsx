import useSettingsContext from '../../hooks/useSettingsContext'
import { renderWithProviders } from '../../../../setupTests'
import Sidebar from './Sidebar'
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import setSidebarVisibility from '../../context/actions/settings/setSidebarVisibility'
import useGetSearchFieldLocationWeather from './components/SidebarSearchField/hooks/useGetSearchFieldLocationWeather'
import useGetGeolocationWeather from './components/SidebarLocations/components/CurrentLocationCard/hooks/useGetGeolocationWeather'
import useGetSavedLocationWeather from './components/SidebarLocations/components/SavedLocationCard/hooks/useGetSavedLocationWeather'

jest.mock('../../hooks/useSettingsContext')
const useSettingsContextMock = useSettingsContext as jest.Mock
const dispatcherMock = jest.fn()

jest.mock(
  './components/SidebarLocations/components/CurrentLocationCard/hooks/useGetGeolocationWeather'
)

jest.mock(
  './components/SidebarLocations/components/SavedLocationCard/hooks/useGetSavedLocationWeather'
)

jest.mock(
  './components/SidebarSearchField/hooks/useGetSearchFieldLocationWeather'
)

const useGetGeolocationWeatherMock = useGetGeolocationWeather as jest.Mock
const useGetSavedLocationWeatherMock = useGetSavedLocationWeather as jest.Mock

const useGetSearchFieldLocationWeatherMock =
  useGetSearchFieldLocationWeather as jest.Mock

describe('Sidebar', () => {
  beforeEach(() => {
    useSettingsContextMock.mockReturnValue({
      state: { isSidebarOpen: true, savedLocations: ['New York'] },
      dispatcher: dispatcherMock
    })

    useGetGeolocationWeatherMock.mockReturnValue({
      isLoading: true,
      hasError: false,
      currentWeather: null
    })

    useGetSavedLocationWeatherMock.mockReturnValue({
      isLoading: true,
      hasError: false,
      currentWeather: null
    })

    useGetSearchFieldLocationWeatherMock.mockReturnValue({
      isLoading: true,
      hasError: false,
      currentWeather: null
    })
  })

  it('should match the snapshot', () => {
    const view = renderWithProviders(
      <Sidebar onSettingsButtonClick={jest.fn()} />
    )

    expect(view.asFragment()).toMatchSnapshot()
  })

  it('should call the dispatcher with the correct action on menu button click', async () => {
    renderWithProviders(<Sidebar onSettingsButtonClick={jest.fn()} />)

    const menuButtonIcon = screen.getByTestId('MenuIcon')
    expect(menuButtonIcon).toBeInTheDocument()

    await userEvent.click(menuButtonIcon)
    expect(dispatcherMock).toHaveBeenCalledWith(setSidebarVisibility(false))
  })

  it('should run the provided settings button click handler on settings button click', async () => {
    const onSettingsButtonClickMock = jest.fn()

    renderWithProviders(
      <Sidebar onSettingsButtonClick={onSettingsButtonClickMock} />
    )

    const settingsButtonIcon = screen.getByTestId('SettingsIcon')
    expect(settingsButtonIcon).toBeInTheDocument()

    await userEvent.click(settingsButtonIcon)
    expect(onSettingsButtonClickMock).toHaveBeenCalled()
  })
})
