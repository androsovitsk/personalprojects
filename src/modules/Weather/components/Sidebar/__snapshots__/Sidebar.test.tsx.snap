// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiDrawer-root MuiDrawer-docked css-qoiqkb-MuiDrawer-docked"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiDrawer-paper MuiDrawer-paperAnchorLeft MuiDrawer-paperAnchorDockedLeft css-12i7wg6-MuiPaper-root-MuiDrawer-paper"
    >
      <div
        class="MuiGrid2-root MuiGrid2-container MuiGrid2-direction-xs-row css-am7og4-MuiGrid2-root"
      >
        <div
          class="MuiGrid2-root MuiGrid2-direction-xs-row css-27y2fm-MuiGrid2-root"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="MenuIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiGrid2-root MuiGrid2-direction-xs-row css-27y2fm-MuiGrid2-root"
        >
          <button
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SettingsIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61l-2.01-1.58zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <h4
        class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
      >
        Weather
      </h4>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          Type in the name of a city...
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r0:"
            type="text"
            value=""
          />
          <div
            class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
              />
            </svg>
          </div>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Type in the name of a city...
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <hr
        class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
      />
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-58ftfq-MuiSkeleton-root"
            style="height: 1rem;"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-58ftfq-MuiSkeleton-root"
            style="height: 1rem;"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-1sqh8dc-MuiSkeleton-root"
            style="width: 66%; height: 1rem;"
          />
        </div>
      </div>
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-1v5z4dq-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-58ftfq-MuiSkeleton-root"
            style="height: 1rem;"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-58ftfq-MuiSkeleton-root"
            style="height: 1rem;"
          />
          <span
            class="MuiSkeleton-root MuiSkeleton-rectangular MuiSkeleton-pulse css-1sqh8dc-MuiSkeleton-root"
            style="width: 66%; height: 1rem;"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
